File: basic_dsa_three_hundred_seventy_six.java
/**
 * Basic DSA Problems - Wiggle Subsequence (LeetCode 376)
 * Longest subsequence where diff alternates positive/negative
 */
public class basic_dsa_three_hundred_seventy_six {

    // Problem: Longest wiggle subsequence (alternating < > < > or > < > <).

    /**
     * Greedy: track up/down length. When trend flips, extend.
     * Time O(n), Space O(1)
     */
    public static int wiggleMaxLength(int[] nums) {
        if (nums.length < 2) return nums.length;
        int up = 1, down = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > nums[i - 1]) up = down + 1;
            else if (nums[i] < nums[i - 1]) down = up + 1;
        }
        return Math.max(up, down);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Wiggle Subsequence");
        System.out.println("=".repeat(60));

        System.out.println("[1,7,4,9,2,5] -> " + wiggleMaxLength(new int[]{1,7,4,9,2,5}));
        System.out.println("[1,17,5,10,13,15,10,5,16,8] -> " + wiggleMaxLength(new int[]{1,17,5,10,13,15,10,5,16,8}));
    }
}

File: basic_dsa_three_hundred_seventy_six.cpp
/**
 * Basic DSA - Longest Harmonious Subsequence (LeetCode 594)
 * Subsequence where max - min == 1; maximize length.
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

int findLHS(std::vector<int>& nums) {
    std::unordered_map<int, int> cnt;
    for (int x : nums)
        ++cnt[x];
    int best = 0;
    for (const auto& p : cnt) {
        int v = p.first;
        auto it = cnt.find(v + 1);
        if (it != cnt.end())
            best = std::max(best, p.second + it->second);
    }
    return best;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Longest Harmonious Subsequence\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {1, 3, 2, 2, 5, 2, 3, 7};
    std::cout << "[1,3,2,2,5,2,3,7] -> " << findLHS(v1) << "\n";

    std::vector<int> v2 = {1, 1, 1, 1};
    std::cout << "[1,1,1,1] -> " << findLHS(v2) << "\n";
}
